File: conformance/tests/property-initializers/property-initializers.runar.zig
const runar = @import("runar");

pub const PropertyInitializers = struct {
    pub const Contract = runar.StatefulSmartContract;

    count: i64 = 0,
    maxCount: i64,
    active: runar.Readonly(bool) = true,

    pub fn init(maxCount: i64) PropertyInitializers {
        return .{ .maxCount = maxCount };
    }

    pub fn increment(self: *PropertyInitializers, amount: i64) void {
        runar.assert(self.active);
        self.count = self.count + amount;
        runar.assert(self.count <= self.maxCount);
    }

    pub fn reset(self: *PropertyInitializers) void {
        self.count = 0;
    }
};

File: conformance/tests/property-initializers/property-initializers.runar.go
//go:build ignore

package contract

import "runar"

type PropertyInitializers struct {
	runar.StatefulSmartContract
	Count    runar.Int
	MaxCount runar.Int `runar:"readonly"`
	Active   runar.Bool `runar:"readonly"`
}

func (c *PropertyInitializers) init() {
	c.Count = 0
	c.Active = true
}

func (c *PropertyInitializers) Increment(amount runar.Int) {
	runar.Assert(c.Active)
	c.Count = c.Count + amount
	runar.Assert(c.Count <= c.MaxCount)
}

func (c *PropertyInitializers) Reset() {
	c.Count = 0
}
